src/dispatch/builder.rs

```diff
@@ -177,6 +177,84 @@ impl<'a, 'b> DispatcherBuilder<'a, 'b> {
+    /// conciseness.
+    ///
+    /// This function accepts the `DispatcherBuilder` as parameter, and the type
+    /// of the `System` that will drive the execution of the internal
+    /// dispatcher.
+    ///
+    /// Note that depending on the dependencies of the SubSystems the Batch
+    /// can run in parallel with other Systems.
+    /// In addition the Sub Systems can run in parallel within the Batch.
+    ///
+    /// The `Dispatcher` created for this `Batch` is completelly separate,
+    /// from the parent `Dispatcher`.
+    /// This mean that the dependencies, the `System` names, etc.. specified on
+    /// the `Batch` `Dispatcher` are not visible on the parent, and is not
+    /// allowed to specify cross dependencies.
+    pub fn with_batch<T>(
+        mut self,
+        dispatcher_builder: DispatcherBuilder<'a, 'b>,
+        name: &str,
+        dep: &[&str],
+    ) -> Self
+    where
+        T: for<'c> System<'c> + BatchController<'a, 'b> + Send + 'a,
+    {
+        self.add_batch::<T>(dispatcher_builder, name, dep);
+
+        self
+    }
+
+    /// The `Batch` is a `System` which contains a `Dispatcher`.
+    /// By wrapping a `Dispatcher` inside a system, we can control the execution
+    /// of a whole group of system, without sacrificing parallelism or
+    /// conciseness.
+    ///
+    /// This function accepts the `DispatcherBuilder` as parameter, and the type
+    /// of the `System` that will drive the execution of the internal
+    /// dispatcher.
+    ///
+    /// Note that depending on the dependencies of the SubSystems the Batch
+    /// can run in parallel with other Systems.
+    /// In addition the Sub Systems can run in parallel within the Batch.
+    ///
+    /// The `Dispatcher` created for this `Batch` is completelly separate,
+    /// from the parent `Dispatcher`.
+    /// This mean that the dependencies, the `System` names, etc.. specified on
+    /// the `Batch` `Dispatcher` are not visible on the parent, and is not
+    /// allowed to specify cross dependencies.
+    pub fn add_batch<T>(
+        &mut self,
+        mut dispatcher_builder: DispatcherBuilder<'a, 'b>,
+        name: &str,
+        dep: &[&str],
+    ) where
+        T: for<'c> System<'c> + BatchController<'a, 'b> + Send + 'a,
+    {
+        dispatcher_builder.thread_pool = self.thread_pool.clone();
+
+        let mut reads = dispatcher_builder.stages_builder.fetch_all_reads();
+        reads.extend(<T::BatchSystemData as SystemData>::reads());
+        reads.sort();
+        reads.dedup();
+
+        let mut writes = dispatcher_builder.stages_builder.fetch_all_writes();
+        writes.extend(<T::BatchSystemData as SystemData>::reads());
+        writes.sort();
+        writes.dedup();
+
+        let accessor = BatchAccessor::new(reads, writes);
+        let dispatcher = dispatcher_builder.build();
+
+        let batch_system = T::create(accessor, dispatcher);
+
+        self.add(batch_system, name, dep);
```

Comment:
Because you can construct `BatchAccessor`, call `create` from the trait and then manually call `self.add`, implementing `Send` batch controller is not safe.

Reply:
Make sense